feat: procedures to copy data

hackaugusto opened the pull request:
These procedures will be useful to copy data to setup the transaction prelude. The prelude needs to copy data from stack to memory, and from the tape to memory (and check the commitment).

I'm opening this to get some feedback on how to test it, suggestions appreciated :). I also will add a procedure to copy data with prefix-length for the notes data.

Related issue #14

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+# assert_eqw(A: Word, B: Word)
+#
+# Asserts that the top two words in the stack are equal
```
For doc comments we should use `#!` rather than just `#` (this way, doc comments get included into the AST). 

Also, we usually describe the initial and the final state of the stack, and sometimes, for simple procedures like this, it might be useful to include a cycle count.

For this procedure, I'd probably have the following description:
```
#! Asserts that the top two words on the stack are equal.
#!
#! Stack transition look like so:
#! [A, B, ...] -> [...]
#!
#! This procedure requires 11 cycles.
```

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+# Moves the `words` words from the advice tape to memory, starting at
+# `writePtr` address, the hash of the copied data is left on the stack for
+# commitment checking.
+proc.move_from_tape_to_memory.0
```
Procedures intended for external consumption should be marked with `export` rather than `proc` keywords. So, for example, this one could be:
```
export.move_from_tape_to_memory
```

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+  movup.13 drop
+end
+
+# move_from_tape_to_memory(words: Felt, writePtr: Address) -> [RpoHash, writePtr']
```
nit: I'd use snake-case for parameter names. For example, `write_ptr` instead of `wirtePtr`.

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+  exec.assert_eqw
+end
+
+# memcopy(readPtr: Address, writePtr: Address, words: Felt)
+#
+# Copies `words` words from `readPtr` to `writePtr`.
+proc.memcopy.0
```
This procedure could probably go into the standard library under `std::mem::memcopy`. In the future, we might even introduce an instruction to perform this in a single VM cycle.

hackaugusto replied:
That is one thing that I had in mind while writing these procedures. I had two goals:

1. write reusable procedures
2. make it easy to change our format (add/remove fields), That is why everything is done with a loop, instead of a manually unrolled loop.

Let me know if I should move these to a PR to `miden-vm`.

bobbinth replied:
Yes, let's make a PR against `miden-vm`, but only for this procedure for now.

hackaugusto replied:
PR is open: https://github.com/0xPolygonMiden/miden-vm/pull/668

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+
+  # LOOP: [readPtr, writePtr, words, ...]
+  # while(words>0)
+  dup.3 neq.0 while.true
+    # prepare read
+    # stack: [readPtr, 0, 0, 0, 0, readPtr, writePtr, words, ...]
+    padw dup.4
+
+    # perform read
+    # stack: [A, readPtr, writePtr, words, ...]
+    mem_loadw
+
+    # prepare write
+    # stack: [writePtr, A, readPtr, writePtr, words, ...]
+    dup.5
+
+    # peform write
+    # stack: [A, readPtr, writePtr, words, ...]
+    mem_storew
+
+    # clean stack
+    # stack: [readPtr, writePtr, words, ...]
+    dropw
+
+    # update counters
+    # LOOP: [readPtr+1, writePtr+1, words-1, ...]
+    inc.1 movup.2 sub.1 movup.2 inc.1 movup.2
+
+    dup.3 neq.0 # while(words>0)
+  end
```
A couple of notes here:

1. I'd avoid pushing and popping values here as much as possible - this will contribute a lot to the cycle count. Instead, we can use 4 elements of the stack as a buffer into which we first `mem_loadw` and then `mem_storew`. We don't need to use `padw` and `dropw` in between these.
2. For the counters, it might be more efficient to keep the target number and count up until the count reaches the target number (rather than counting to 0). This is because incrementing by 1 is cheap (1 VM cycles) but subtracting 1 is not as cheap (3 VM cycles). Also, I don't think `inc.1` is a valid instruction - you can use `add.1` instead.

hackaugusto replied:
just for reference, this discussion continued on the miden-vm PR https://github.com/0xPolygonMiden/miden-vm/pull/668

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+#
+# Moves the `words * 4` elements from the stack to memory, starting at
+# `writePtr` address
+proc.move_from_stack_to_memory.0
```
This is a cool procedure, but I'm not sure which context it will be useful in.

hackaugusto replied:
It is to copy the global inputs from the stack to the memory. Since that section is not fully spec'ed, I opted to write a small procedure to copy the data, so that I can just update the counter if we do some changes. My idea here was to optimize for flexibility, so that we can iterate over the format quicker, instead of optimize for efficiency in number of cycles.

bobbinth commented on kernel/copy.masm:
```diff
@@ -0,0 +1,256 @@
+  # stack: [needs_padding, A, B, C, writePtr, ...]
+  movup.13
+
+  # if(needs_padding)
+  if.true 
+    # Rescue Prime Optimized uses overwrite mode. Drop `A`.
+    # stack: [B, C, writePtr, ...]
+    dropw
+
+    # Overwrite the `B` with the new data
+    # stack: [B', C, writePtr, ...]
+    adv_loadw
+
+    # - get the memory address that B' should be saved to
+    # - update the writePtr to point to the next address
+    # stack: [writePtr, B', writePtr+1, C, ...]
+    movup.8 dup.0 add.1 movdn.5
+
+    # save data to memory
+    # stack: [B', writePtr+1, C, ...]
+    mem_storew
+
+    # Fix writePtr position
+    # stack: [B', C, writePtr+1, ...]
+    movup.4 movdn.8
+
+    # Push padding word
+    # stack: [A, B', C, writePtr+1, ...]
+    push.1.0.0.0
+
+    # Run RPO permutation
+    # stack: [A', B', C', writePtr+1, ...]
+    rpperm
+  end
```
I'll need to think a bit more about how we want to handle padding here. It might be simpler to require the padding be done correctly outside of the VM rather than doing in the VM (i.e., the advice provider will contain padded data already). I don't believe this will be a security risk - but need to think it through.

Another thing: currently we are specifying the number of words to read, but shouldn't it be the number of elements? Or if we wanted to make this limited to words, maybe the name of the procedure should be `move_word_from_tape_to_memory`?

hackaugusto replied:
> It might be simpler to require the padding be done correctly outside of the VM rather than doing in the VM

No problem, the code should handle both cases. At least for the moment I was optimizing for flexibility, that is why this is built-in. I think once we have the data structures properly defined we can write custom code that doesn't use loops to load the data.

> currently we are specifying the number of words to read, but shouldn't it be the number of elements?

I wrote the procedures with words because that is what we have on the issue spec. I can rename it and add another version that works on elements too.

bobbinth reviewed (commented):
Thank you! Looks good! I left a few comments inline.

In terms of how to test, we could probably set up a testing structure similar to how we test procedures from the standard library (see [here](https://github.com/0xPolygonMiden/miden-vm/blob/main/miden/tests/integration/stdlib/math/u64_mod.rs)).

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  # stack: [write_ptr, words, 0, 0, ...]
+  push.0 movdn.2 push.0 movdn.2
+  
+  # whiel(words!0)
```
nit: typo on "while", and I'd probably use `!=` as well

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+    # stack: [write_ptr+1, words-1, 0, 0, ...]
+    add.1 swap sub.1 swap
+
+     # whiel(words!0)
```
same nit as above

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+export.move_from_stack_to_memory.0
+  
+  # align the top of the stack in a word boundary, so that stack data can be
+  # moved with word operations. Leave write_ptr on the top so that increamenting
```
nit: typo in "incrementing"

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  dropw swapw dropw
+end
+
+#! move_from_tape_to_memory_with_commitment(words: felt, write_ptr: address, commitment: commitment) -> [write_ptr']
+#!
+#! Moves the `words` words from the advice tape to memory, starting at
+#! `write_ptr` address, at the end asserts the commitment data was copied
+#!
+#! input: [words, write_ptr, commitment, ...]
+#! output: [write_ptr', ...]
+export.move_from_tape_to_memory_with_commitment.0
```
nit: should we include the cycle count?

hackaugusto replied:
👍 added to all of the procedures

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  drop drop drop
+end
+
+#! move_from_stack_to_memory(write_ptr: address, words: felt)
+#!
+#! Moves the `words * 4` elements from the stack to memory, starting at
+#! `write_ptr` address
+#
+#! input: [write_ptr, words, ...]
+#! output: [...]
+export.move_from_stack_to_memory.0
```
nit: should we include the cycle count?

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  movup.13 drop
+end
+
+#! move_from_tape_to_memory(words: felt, write_ptr: address) -> [rpo_hash, write_ptr']
+#!
+#! Moves the `words` words from the advice tape to memory, starting at
+#! `write_ptr` address, the hash of the copied data is left on the stack for
+#! commitment checking.
+#!
+#! input: [words, write_ptr, ...]
+#! output: [A, B, C, write_ptr', ...]
+export.move_from_tape_to_memory.0
```
should we include a cycle count?

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  # stack: [is_odd, words, write_ptr, needs_padding, ...]
+  dup movdn.3
+
+  # Make sure `words` is event. This is a requirement for the
```
nit: typo in "even" here and a few lines below as well.

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  # stack: [words_even, write_ptr, needs_padding, ...]
+  sub
+
+  # Prepare the capacity word. For rescue prime optimized the first element is
+  # set to `1` when padding is used and `0` otherwse, this is determined by the
+  # `needs_padding` flag, so just copy it.
+  dup.1 push.0.0.0
```
Doesn't this copy the `write_ptr`?

hackaugusto replied:
I'm going to mark this as a draft, I need to write tests for these.

grjte commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+  # `needs_padding` flag, so just copy it.
+  dup.1 push.0.0.0
+
+  # set initial state
+  # stack: [A, B, C, write_ptr, words_even, needs_padding, ...]
+  padw padw
```
Is this stack order correct? I don't think `write_ptr` and `words_even` got swapped

grjte reviewed (dismissed):
Thanks @hackaugusto! I'm not sure what the status of this is now that memcopy has been moved out. I've done a partial review for now

hackaugusto commented on kernel/copy.masm:
```diff
@@ -0,0 +1,265 @@
+    # stack: [[write_ptr', words, x, x], A', B', C', ...]
+    swapw.3
+
+    # stack: [[words, write_ptr', x, x], A', B', C', ...]
+    movup.1
```
thanks! updated :) I have to add tests for these instructions

hackaugusto commented on objects/Cargo.toml:
```diff
@@ -19,6 +19,6 @@ default = ["std"]
-assembly = {  package = "miden-assembly", git = "https://github.com/0xPolygonMiden/miden-vm", branch = "next", default-features = false }
-crypto = {  package = "miden-crypto", git = "https://github.com/0xPolygonMiden/crypto", branch = "next", default-features = false }
-miden-core = {  package = "miden-core", git = "https://github.com/0xPolygonMiden/miden-vm", branch = "next", default-features = false }
+assembly = {  package = "miden-assembly", git = "https://github.com/0xPolygonMiden/miden-vm", default-features = false, rev = "914059c4e72965acbae78ca285b775ae30408d4b"  }
```
This can be updated once https://github.com/0xPolygonMiden/miden-vm/pull/687 is merged (or we find an alternative to that)

hackaugusto commented on tx_kernel/asm/copy.masm:
```diff
@@ -0,0 +1,229 @@
+#! assert_eqw(A: Word, B: Word)
```
@bobbinth I used this style on my example, but you mentioned on the review of `memcopy` that you find it confusing, let me know if I should follow the same style here.

hackaugusto commented on tx_kernel/build.rs:
```diff
@@ -0,0 +1,28 @@
+/// Read and parse the contents from `./asm` into a `LibraryContents` struct, serializing it into
+/// `assets` folder under `std` namespace.
+#[cfg(not(feature = "docs-rs"))]
+fn main() -> io::Result<()> {
```
I removed the documentation generation since there is not much going. Let me know if I should add back (and were should we put the md generator)

frisitano commented:
Whats the status on this?  Are we intending on merging?  I want to make a start on #16 and the repo structure we have here could be useful?

hackaugusto commented:
@frisitano this is waiting for a review on https://github.com/0xPolygonMiden/miden-vm/pull/687 (latest status was to add a flag to the `Processor` to hide its internal state)

frisitano commented:
> @frisitano this is waiting for a review on [0xPolygonMiden/miden-vm#687](https://github.com/0xPolygonMiden/miden-vm/pull/687) (latest status was to add a flag to the `Processor` to hide its internal state)

Ok thanks for the update.  Do you mind if I add the flag?

hackaugusto commented:
> Ok thanks for the update. Do you mind if I add the flag?

not at all :)

frisitano commented:
> > Ok thanks for the update. Do you mind if I add the flag?
> 
> not at all :)

I created a PR for the `internals` flag in `miden-vm` (there had been quite a few changes so it was easier to just branch from next).  I also updated this PR accordingly - see https://github.com/0xPolygonMiden/miden-base/compare/595f327ce2123918f34a9201a4a1d77f24c589ae..1861975799b836629995e7e2951baee6b85d9c32.

bobbinth commented:
I guess there are two parts in this PR:

1. Setting up structure of the tx kernel library.
2. Adding some utility methods which we might need to use in the future.

Regarding the first one - I'd actually prefer a different structure. I think we should have a single `miden` library in Miden assembly which will contain tx kernel and other things (e.g., kernel procedures described [here](https://github.com/0xPolygonMiden/miden-base/issues/3#issuecomment-1396178442) would go there too). The crate name for this could be `miden-lib` or something like that.

Tx kernel prologue, epilogue etc. could go into this library under namespace `miden::sat::prologue`, `miden::sat::epilogue`, etc.

Regarding the second part: I'm actually not yet sure all of these procedures are needed, and some of them may make more sense as assembly instructions (e.g., `assert_eqw` and `is_odd` are good candidates to be added as assembly instructions).

bobbinth commented:
> I want to make a start on #16 and the repo structure we have here could be useful?

@frisitano - I wonder if a better place to start would be #14.

hackaugusto commented:
- Updated the code to use the latest release which has the `internals` flag.
- Moved the code to `miden_lib`

> Regarding the second part: I'm actually not yet sure all of these procedures are needed, and some of them may make more sense as assembly instructions (e.g., assert_eqw and is_odd are good candidates to be added as assembly instructions).

@bobbinth which procedures should be made into instructions or removed?

bobbinth commented:
I think `is_odd` could definitely be an instruction, and probably `assert_eqw` as well.

I'm not sure about the use case for `move_words_from_stack_to_memory` - usually, when we move words from stack to memory, it is just a handful of words an we know exactly how many words to move - so, generic implementation might be an overkill.

For `move_words_from_tape_to_memory_with_commitment`, I think we should probably use `adv_pipe` instruction - it should be much more efficient than what we have implemented. Same probably goes for `move_words_from_tape_to_memory` when the number of words is more than a few. And I'll probably need to think through of the best way to handle even vs. odd number of words.

hackaugusto commented:
> I think is_odd could definitely be an instruction, and probably assert_eqw as well.

Opened the first PR for `assert_eqw` https://github.com/0xPolygonMiden/miden-vm/pull/734

> I'm not sure about the use case for move_words_from_stack_to_memory - usually, when we move words from stack to memory, it is just a handful of words an we know exactly how many words to move - so, generic implementation might be an overkill.

deleted the procedure

> For move_words_from_tape_to_memory_with_commitment, I think we should probably use adv_pipe instruction - it should be much more efficient than what we have implemented. Same probably goes for move_words_from_tape_to_memory when the number of words is more than a few. And I'll probably need to think through of the best way to handle even vs. odd number of words.

The procedures are using `adv_pipe` and handle padding. Anything missing?
